broker/lib/fabrik/DirectorManager.js

```diff
@@ -177,7 +189,179 @@ class DirectorManager extends BaseManager {
+          maxWorkers = targetDirectorConfig.policies.user[action].max_workers;
+        }
+        if (currentTasks < maxWorkers) {
+          runOutput.shouldRunNow = true;
+          return Promise.resolve(runOutput);
+        }
+        runOutput.shouldRunNow = false;
+        return Promise.resolve(runOutput);
+      }).catch(err => {
+        logger.error('Error connecting to BOSH director > could not fetch current tasks', err);
+        //in case the director request returns an error, we queue it to avoid user experience issues
+        runOutput.directorError = true;
+        return Promise.resolve(runOutput);
+      });
+  }
+
+  removeCachedTask(serviceInstanceId) {
+    return boshOperationCache.deleteBoshTask(serviceInstanceId);
+  }
+
+  getCurrentOperationState(serviceInstanceId) {
+    let output = {
+      'cached': false,
+      'task_id': null
+    };
+
+    return Promise.all([boshOperationCache.containsServiceInstance(serviceInstanceId), boshOperationCache.getBoshTask(serviceInstanceId)])
+      .spread((cached, taskId) => {
+        output.cached = cached;
+        output.task_id = taskId;
+      })
+      .return(output);
+  }
+
   createOrUpdateDeployment(deploymentName, params, args) {
+    logger.info(`Checking rate limits against bosh for deployment `);
+    const previousValues = _.get(params, 'previous_values');
+    const action = _.isPlainObject(previousValues) ? CONST.OPERATION_TYPE.UPDATE : CONST.OPERATION_TYPE.CREATE;
+    const scheduled = _.get(params, 'scheduled') || false;
+    const dbUpdate = _.get(params, '_serviceFabrikDbUpdate') || false;
+    const policyApplicable = config.enable_bosh_rate_limit;
+    _.omit(params, 'scheduled');
+    _.omit(params, '_serviceFabrikDbUpdate');
+
+    const decisionMaker = {
+      'shouldRunNow': false,
+      'cached': false,
+      'dbUpdate': false,
+      'policyApplicable': config.enable_bosh_rate_limit,
+      'directorError': false
+    };
+
+    return this.executePolicy(scheduled, action, deploymentName, dbUpdate, policyApplicable)
+      .then(checkResults => {
+        if (checkResults.directorError) {
+          decisionMaker.directorError = true;
+          return Promise.resolve(true);
+        }
+        const shouldRunNow = checkResults.shouldRunNow;
+
+        if (!decisionMaker.policyApplicable) {
+          return Promise.resolve();
+        }
+        if (dbUpdate) {
+          decisionMaker.shouldRunNow = true;
+          decisionMaker.dbUpdate = true;
+          return Promise.resolve();
+        }
+        if (shouldRunNow) {
+          decisionMaker.shouldRunNow = shouldRunNow;
+          return boshOperationCache.containsDeployment(deploymentName).then(inCache => {
+            if (inCache) {
+              decisionMaker.cached = true;
+            }
+            return Promise.resolve();
+          });
+        }
+        if (!scheduled) {
+          return Promise.resolve(true);
+        }
+        throw new DeploymentDelayed(deploymentName);
```

Comment:
why is this exception being thrown in case of scheduled and should not be runNow? for user tasks why is exception not being thrown?

Reply:
For user tasks, the exception is thrown later on once it's saved in ETCD. For scheduled tasks, we do not store in etcd because the scheduler will pick it up for processing later on - the exception is just thrown in the case of scheduled ops